source/agora/consensus/BlockRewardCalc.d

```diff
@@ -0,0 +1,651 @@
+/*******************************************************************************
+
+    Contains functions to calculate the block reward.
+
+    1. We will only distribute block rewards in every X blocks
+       as opposed to distributing it in every block. This was originally meant
+       to reduce blockchain size.
+    2. We will delay the distribution of the block rewards by Y blocks, so for
+       blocks A .. B we will distribute the rewards at block B + Y.
+       This is because we want to give enough time validators to share
+       block signatures which is the basis of the block reward.
+    3. Each validator will be rewarded based on how many blocks they signed
+       and the reward will NOT be propotional to how much coin they staked.
+    4  The block reward is divided between the validators and the Commons Budget.
+    5. The total block reward for a particular height is fixed,
+       however this doesn't mean that the total validator reward is fixed.
+       The total validator reward is based on the total number of signatures
+       on the blocks in question. The total validator reward is a
+       nonlinear function of the the total signatures,
+       so 2 times more signatures might increase the total validator reward by 4.
+       This is to encourage validators to share signatures.
+       The coins remaing is  'fixed block reward at height X' -
+       'total validator reward' will go to the Commons Budget.
+    6. The part of the total validator reward that cannot be didived equally
+       between the validators will also go to the Commons Budget
+
+    Copyright:
+        Copyright (c) 2021 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.BlockRewardCalc;
+
+import agora.common.Amount;
+import agora.common.Types;
+import agora.consensus.data.Params;
+import agora.crypto.ECC : Point;
+import agora.utils.Log;
+
+import std.algorithm : min;
+import std.array;
+import std.math;
+import std.range : repeat;
+import std.typecons : Tuple, tuple;
+
+import core.time;
+
+mixin AddLogger!();
+
+///
+public class BlockRewardCalc
+{
+
+    ///
+    public this (immutable(ConsensusParams) consensus_params) @safe @nogc pure nothrow
+    {
+        this.consensus_params = consensus_params;
+    }
+
+    ///
+    private alias CurrentPeriodTup = Tuple!(ulong, "period", bool, "is_period_end");
```

Comment:
A `struct` would also work :)

Reply:
Just wondering if struct would have any benefit over Tuples or if we should refine our coding guidelines. Tuples have some nice methods like expand, translate... Any suggestion on when to use one over the other?